xarray/tests/test_backends.py

```diff
@@ -617,65 +651,52 @@ def test_roundtrip_string_with_fill_value_nchar(self):
             with self.roundtrip(original) as actual:
                 assert_identical(expected, actual)
 
-    def test_unsigned_roundtrip_mask_and_scale(self):
-        decoded = create_unsigned_masked_scaled_data()
-        encoded = create_encoded_unsigned_masked_scaled_data()
+    @pytest.mark.parametrize(
+        'decoded_fn, encoded_fn',
+        [(create_unsigned_masked_scaled_data,
+          create_encoded_unsigned_masked_scaled_data),
+         pytest.param(create_bad_unsigned_masked_scaled_data,
+                      create_bad_encoded_unsigned_masked_scaled_data,
+                      marks=pytest.mark.xfail(reason="Bad _Unsigned attribute.")),
+         (create_signed_masked_scaled_data,
+          create_encoded_signed_masked_scaled_data),
+         (create_masked_and_scaled_data,
+          create_encoded_masked_and_scaled_data)])
+    def test_roundtrip_mask_and_scale(self, decoded_fn, encoded_fn):
+        decoded = decoded_fn()
+        encoded = encoded_fn()
+
         with self.roundtrip(decoded) as actual:
             for k in decoded.variables:
                 assert (decoded.variables[k].dtype ==
                         actual.variables[k].dtype)
             assert_allclose(decoded, actual, decode_bytes=False)
+
         with self.roundtrip(decoded,
                             open_kwargs=dict(decode_cf=False)) as actual:
+            # TODO: this assumes that all roundtrips will first
+            # encode.  Is that something we want to test for?
             for k in encoded.variables:
                 assert (encoded.variables[k].dtype ==
```

Comment:
`(create_masked_and_scaled_data, create_encoded_masked_and_scaled_data)` fails these dtype tests. I don't know what's happening. Help?

Reply:
Example error:

```
               assert (encoded.variables[k].dtype ==
                        actual.variables[k].dtype)
E               AssertionError: assert dtype('int64') == dtype('int16')
E                +  where dtype('int64') = <xarray.Variable (t: 5)>\narray([-1, -1,  0,  1,  2])\nAttributes:\n    _FillValue:    -1\n    add_offset:    10\n    scale_factor:  0.1.dtype
E                +  and   dtype('int16') = <xarray.Variable (t: 5)>\ndask.array<shape=(5,), dtype=int16, chunksize=(5,)>\nAttributes:\n    add_offset:    10\n    scale_factor:  0.10000000149011612\n    _FillValue:    -1.dtype
```

Reply:
This may be as easy as setting the dtype to `'i2` for the array in create_encoded_masked_and_scaled_data. I think this was a simple oversight to not set the dtype, but clearly that's necessary to match the specified encoding.